### DISCUSSION TITLE
Disable browser open/refresh upon `retype watch`

### DISCUSSION THREAD

**Discussion by rdewit** [retypeapp/retype#323] [Q&A]
Currently retype opens/refreshes the browser window upon `retype watch` in a similar way that Create React App does this. In my case, this is a nuisance as my default browser (Firefox) is not my dev browser (Google Chrome) and Firefox opens a new tab every time `retype watch` is called.

Would it be possible to allow the dev to disable this feature via an env var (like CRA's `BROWSER=none`)?

**Comment by geoffreymcgill** [comment]
Hello. You can run [`retype build`](https://retype.com/guides/cli/#retype-build) instead, which will just generate the website and you can decide when to start a web server. 

You can start your own web server, or call [`retype run`](https://retype.com/guides/cli/#retype-run) when you're ready.

The `retype watch` command is essentially a short cut for running the following commands in the following sequence:
1. `retype init`
2. `retype build`
3. `retype run`
4. Open in a browser

Hope this helps.

**Reply by rdewit** [reply]
Hi @geoffreymcgill , thank you for the very quick response. I may be missing something, but `retype run` does not seem to watch for changes in the content and refreshing the page, and I want to maintain that functionality.

**Reply by rdewit** [reply]
My issue is with a browser tab getting opened when I call `retype watch`. The refresh upon content changes is exactly what I want.

**Reply by geoffreymcgill** [reply]
> but retype run does not seem to watch for changes in the content and refreshing the page

Correct. 

The intention with `retype watch` is that you should theoretically only need to call the command once. Or at least once per development session. 

What is the scenario for which you need to call `retype watch` multiple times?

Disabling the browser open functionality would be possible, although we would need to build in an extra command argument or an env var as you suggested. I will have to investigate.

**Reply by rdewit** [reply]
> What is the scenario for which you need to call `retype watch` multiple times?
> 
> Disabling the browser open functionality would be possible, although we would need to build in an extra command argument or an env var as you suggested. I will have to investigate.

I initially stopped/started `retype watch` multiple times because I didn't realize that issue #324 only required a manual browser refresh when changing a CSS file.

However, I'd like to have control over whether or not a browser window gets opened, especially since it opens the 'wrong' browser window for me.

It would be great if this could be made configurable one way or another! :-)

**Reply by geoffreymcgill** [reply]
> It would be great if this could be made configurable one way or another! :-)

Thank you for the feature request. We will try to come up with some options.

**Reply by rdewit** [reply]
Until then, I'm using a nasty hack (likely only to work on Linux):

`DISPLAY=:none retype watch`

**Reply by DannyBen** [reply]
I am also annoyed by this behavior (which as it seems was added recently).

I am using a headless Ubuntu, and I have a text based browser installed - running retype watch, launches this browser.

The common behavior I saw in other tools that provide watch / build functionality is to have the flag `--open` for people who want to open it in a browser. To me this makes much more sense, since running the command `retype watch` implies that the user wants to build-and-watch, and running `retype watch --open` means "... also open".

Assuming that all `retype watch` users are using a terminal with a browser is a faulty assumption.

Same goes for `retype run`. Now, I cannot enjoy the nice smooth live reload that the retype CLI provides, instead I have to hack my own retype build + file watcher + static file server. 

I would even go a little further, and claim that the `watch` command does not need to start serving, unless requested to do so: `retype watch --serve --open`.

**Reply by geoffreymcgill** [reply]
I agree, the current CLI could use some refinements.

Here are my proposed CLI changes:

1. Rename `retype watch` to `retype start` (or `retype launch`?)
2. Rename `retype run` to `retype serve`
3. Remove the automatic default web browser open from `retype serve` (previously `retype run`)
4. Add `--no-open` to `retype start` (previously `retype watch`) to suppress the automatic browser open

Leave `watch` and `run` would remain as aliases `retype start` and `serve`. All references to `retype watch` and `retype run` would be revised within the documentation and all editable online sources.

I'm kinda partial to `retype launch` instead of `retype start`, although `retype start` is generic and commonly used for this scenario.

Please let me know what you think of the above proposal and if you have any additional changes or renaming suggestions.

Hope this helps.

**Reply by DannyBen** [reply]
Thanks for being open to it. If this is being considered, I would like to map all the "actions" that may be desired to work together (I hope I didn't forget anything).

Also, for me it always helps to consider the `--flags` as an "and" in the sentence: "Build and --watch" - when commands are designed like this, I feel it helps to register on the intuition level.

| Build | Watch-then-build | Serve | Open | Command |
|-------|------------------|-------|------|---------|
| X     |                  |       |      | **`retype build`** |
| X     | X                |       |      | `retype build --watch` |
| X     | X                |       | X    | `retype build --watch --open` |
| X     |                  |       | X    | `retype build --open` |
|       |                  | X     |      | **`retype serve`** |
|       | X                | X     |      | `retype serve --watch` |
|       | X                | X     | X    | `retype serve --watch --open` |
|       |                  | X     | X    | `retype serve --open` |
|       |                  |       | X    | **`retype open`** |


Of course, some people's intuition might be different, and one can argue that `watch` can be a command of its own, which can also be fine, but personally when "watching and building" I always find myself wanting it to first build one time, and then start watching. So: Build then watch.

Also, it would be nice if...

- Long flags will have their short flag version: `--watch, -w` and `--open, -o`
- Commands have a short letter alias: `serve, s`, `build, b` etc.

Lastly:

- I would probably consider avoiding `retype start` - since it can be confusing with more than 2 concepts competing on the same word: Start what? the server? the builder? a new project (`init`)? The browser?
- I would avoid `--no-open`. Instead of being negative, be positive. The open task should not be default anywhere, except a dedicated `open` command. Be considerate of headless-terminal users. People would not complain if `serve` does not open (unless they add `-o`), but on the other hand, if I had to always type `--no-open` I would definitely resort to creating a terminal alias that is shorter to encompass this flag always, which hints that it would be annoying.

**Reply by geoffreymcgill** [reply]
@DannyBen Nice suggestions. Thank you for the feedback.

We're getting close to a solution, although I have a few proposed changes. Comments at the bottom.

### Commands

```
retype init
retype build
retype serve  <-- previously run
retype start  <-- previously watch
```

### Proposed command matrix

| init | build | serve | --watch | --open | Command                       |
|------|-------|-------|---------|--------|-------------------------------|
| X    | X     | X     | X       | X      | `retype start`               |
| X    |       |       |         |        | `retype init`                 |
|      | X     |       |         |        | `retype build`                |
|      | X     |       | X       |        | `retype build --watch`        |
|      |       | X     |         |        | `retype serve`                |
|      |       | X     |        | X      | `retype serve --open` |

`retype start` is the equivalent of running the following sequence of commands:

```cmd
retype init
retype build --watch
retype serve --open
```

Or, in one line:

```
retype init && retype build --watch && retype serve --open
```

Or, in one line with short flags

```
retype init && retype build -w && retype serve -o
```

### Backwards compatibility

Create aliases for the legacy commands which map to the new commands. Functionality of the legacy commands must remain unchanged and no breaking changes to the CLI.

Legacy | New
--- | ---
`retype watch` | `retype start`
`retype run`   | `retype serve`

The legacy commands will be removed from the documentation and any other editable sources.

### New flags

Long | Short
--- | ---
`--watch`   | `-w`
`--open`    | `-o`

### Comments

1. Retype requires a simple command (`retype start`) that can be run on any folder of Markdown files and without any other manual revisions. Retype must be able to build those pages and launch the newly generated website into a browser. This is part of the beauty of Retype and the simplicity for new users. Run one simple command (after installation) and your website is running within seconds.,
2. `retype build` does not require an `--open` flag, as `build` does not start the web server.
3. `retype build --watch` will watch for file changes in the `input` folder and automatically trigger a rebuild.
5. By default, `retype serve` does not open the default web browser. Using the command `retype serve --open` will open the default web browser. 

Let me know what you think.

**Reply by DannyBen** [reply]
Sounds excellent.

- About comment 2: Absolutely right, my mistake.
- About the shortcut `launch` command - although not usable to me personally, having shortcuts is always nice
- About backwards compatibility - I think that if it can be obtained easily, that is great. But I would not spend too much time on it. If a broken backwards compatibility can be fixed on the user side in less than 10 minutes, its fine to require the users to do so in my mind.

**Reply by geoffreymcgill** [reply]
We feel the term `launch` is not ideal. We are proposing to use `retype start` as a replacement for `retype watch`, instead of `retype launch`.

Any thoughts?

**Reply by geoffreymcgill** [reply]
@DannyBen I see you left some feedback on `start` a while back.

> I would probably consider avoiding retype start - since it can be confusing with more than 2 concepts competing on the same word: Start what? the server? the builder? a new project (init)? The browser?

Another option would be repurposing `retype run`, but that would cause a breaking change for anyone currently using `retype run`.

**Reply by DannyBen** [reply]
Well. I am not sure the status, which commands are in still question, and which commands are approved for development.

My thoughts:

1. Avoid ambiguity wherever possible. So `init`, `build` and `serve` - there is no doubt what each of these do.
2. If there are shortcuts - like `watch` == `server --watch`, that is fine - as long as they are not violating the previous rule, if possible.
3. If only `launch` is in question - I am not married to this word either. Personally, I am not even sure this shortcut is needed.
4. I do not think `run` or `start` should be in the command line at all, if there are `serve` and `build`. The reason is that anyone reading the list of commands, might be confused as to what is their purpose.
5. Changing a command line, especially if it is without removing any functionality, is a very minor breaking change. If anything can be fixed in less than 2 minutes, I believe it can even be released in a minor version, let alone in a major version.

That said - for me, it is more important to have granular control over the different operations (serve / watch / build / open), in the shortest and most intuitive manner. Everything else is secondary.

**Reply by geoffreymcgill** [reply]
The current list of commands and flags we are implementing for Retype v3.0 are listed in the following table:

| `init` | `build` | `serve` | `--watch` or `-w` | `--open` or `-o` | Command                       |
|------|-------|-------|---------|--------| :------------------------------- |
| ✅     | ✅     | ✅     | ✅       | ✅      | `retype start`               |
| ✅    |       |       |         |        | `retype init`                 |
|      | ✅     |       |         |        | `retype build`                |
|      | ✅     |       | ✅       |        | `retype build --watch`        |
|      |       | ✅     |         |        | `retype serve`                |
|      |       | ✅     |         | ✅      | `retype serve --open` |

`retype start` is the equivalent of running the following sequence of commands:

```cmd
retype init
retype build --watch
retype serve --open
```

### Backwards compatibility

Legacy | New
--- | ---
`retype watch` | `retype start`
`retype run`   | `retype serve`

### New flags

Long | Short
--- | ---
`--watch`   | `-w`
`--open`    | `-o`

**Reply by DannyBen** [reply]
So the current release's `retype watch` is replaced with `retype serve`, which always watches - yes?
If so, looks great (if not, it needs a `--watch` flag, and then it will be great).

**Reply by geoffreymcgill** [reply]
`retype start` is replacing `retype watch`.

`retype serve` is replacing `retype run`.

I added some more details [above](https://github.com/retypeapp/retype/discussions/323#discussioncomment-5445295).

**Reply by DannyBen** [reply]
Why? Why have `run`, `serve`, `build` and `start`?.... I thought we passed that confusion point.
This makes no sense to me. I think there are three main functions for retype:

1. Init
2. Build
3. Serve

and two flags:

1. `--open` (applicable only for `serve`)
2. `--watch` (applicable for `build` and for `serve` - where in the case of the latter, it can be implicit)

That's it. Simple and intuitive.

If there are strong proponents for having a `start` shortcut, so that in the "Getting Started" tutorial, you can just instruct users to `retype start` - that is fine, although personally I think it is redundant.

> **"Perfection is achieved, not when there is nothing more to add, but when there is nothing left to take away."**
>    *Antoine de Saint-Exupéry*

**Reply by geoffreymcgill** [reply]
As mentioned above, the `retype run` command is being replaced with `retype serve`. There will be only three primary commands:

1. `init`
2. `build`
3. `serve`

and the two flags:

1. `--watch` (only applies to `retype build`)
2. `--open` (only applies to `retype serve`)

We are implementing `retype start` because asking a new non-technical user to input the command `retype init && retype build --watch && retype serve --open` is not exactly user friendly.

**Reply by DannyBen** [reply]
So once again, I don't understand if `retype serve` is also going to refresh the page when I edit the markdown.
For me `retype start` is of no use. It is doing too much (namely the unwanted `--open` and unneeded `--init`).

If `retype serve` is doing an implicit watch, excellent. If not, it should have a `--watch` flag. One of the two. 

Your explanation above of `retype watch` (which is what I use now) being replaced with `retype start` is what confuses me. I expect `retype serve` to be the equivalent of the old `retype watch` minus the init and the open.

I see that the current `retype run` means "serving a static site" - which to me, means "just serve the already built site" - therefore, I feel there is even more place for confusion that can be easily solved:

- `retype serve` - build once and serve the static site, no watching
- `retype serve --watch` - as mentioned above, build, serve, and watch for changes.

**Reply by geoffreymcgill** [reply]
Here is the list of commands and flags we have implemented for Retype v3.0:

| `init` | `build` | `serve` | `--watch` or `-w` | `--open` or `-o` | Command (v3.0) | Previous | 
|:------:|:-------:|:-------:|:---------:|:--------:| :------------------------------- | :--- |
| ✅  | ✅   | ✅  | ✅    | ✅   | `retype start`                | `retype watch` | 
| ✅  |       |       |         |        | `retype init`                   | `retype init` | 
|      | ✅   |       |         |        | `retype build`                | `retype build` |
|      | ✅   |       | ✅    |        | `retype build --watch`  | | 
|      |       | ✅   |         |        | `retype serve`               | `retype run` |
|      |       | ✅   | ✅    |        | `retype serve --watch` | | 
|      |       | ✅   |         | ✅   | `retype serve --open`   | |
|      |       | ✅   | ✅    | ✅   | `retype serve --watch --open` | |

| Command&nbsp;(v3.0)&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;| Description |
| ------------------------------- |  --- |
| `retype start`                | Build and serve the project using one command. | 
| `retype init`                   | Initialize the project with a new `retype.yml` project file. |  
| `retype build`                | Build the project. | 
| `retype build --watch` | Build the project, watch for file changes in the `input`, and rebuild if any changes. | 
| `retype serve`               | Serve the project using the local development only web server. | 
| `retype serve --watch` | Serve the project using the local development only web server and watch for changes in the `output`. If any changes, update all open local web browsers. |
| `retype serve --open`   | Serve the project using the local development only web server and automatically open the default web browser. |
| `retype serve --watch --open` | Serve the project using the local development only web server, watch for changes in the `output` folder, and automatically open the default web browser. Retype will update all open local web browsers if any file changes are detected in the `output` folder. |

The command `retype start` is the equivalent of running the commands in sequence:

```
retype init
retype build --watch
retype serve --watch --open
```

This was the last remaining piece of functionality under development for Retype v3.0. Currently, we are deep testing v3.0. If no critical issues are found, we expect to release v3.0 right away.

Feedback is welcome and changes are possible at this point. Let me know what you think.

**Reply by DannyBen** [reply]
If there is a way for me to install a version with this functionality, I can test as well.

**Reply by geoffreymcgill** [reply]
We are in the process of making a public `v3.0.0-preview` release and it should be available today, as long as the release process doesn't encounter any issues.

**Reply by geoffreymcgill** [reply]
The `v3.0.0-preview` release is now available, see
https://www.npmjs.com/package/retypeapp/v/3.0.0-preview

You should be able to install or update your local install using either of the following NPM commands:

```
# Install Retype v3.0.0-preview release
npm install retypeapp@v3.0.0-preview --global

# or, update if Retype was installed previously
npm update retypeapp@v3.0.0-preview --global
```

If you are using the Retype GitHub Build action, you can update your `retype-action.yml` file with the following, where `@latest` is replaced with `@v3.0.0-preview`.

```yml
    steps:
      - uses: actions/checkout@v2

      - uses: retypeapp/action-build@v3.0.0-preview

      - uses: retypeapp/action-github-pages@v3.0.0-preview
        with:
          update-branch: true
```

Please note, once `v3.0.0` final has been released (not the `-preview`), you should switch back to `@latest`. Using the `@v3.0.0-preview` tag is only for testing the `-preview` release. 

Hope this helps.

**Reply by DannyBen** [reply]
Bug report:

Running `retype serve --watch` does not build on initial run, and does not watch.

For comparison, I tested `retype start` and `retype build --watch` - both work as expected (building and rebuilding on change).

I am using polling mode, if that makes any difference:

```yaml
server:
  host: 0.0.0.0
  port: 3000
  watch: { polling: 3000 }
```

and the entire Retype site source is [here](https://github.com/dannyben/runfile-book).

**Reply by geoffreymcgill** [reply]
> Running `retype serve --watch` does not build on initial run, and does not watch.

This is the expected behaviour. The functionality of `retype serve --watch` is to:
1. Start the web server
2. Serve the website from the `output` folder
3. Watch for changes in the `output` folder
4. Update any connected web browsers if there is a change

The command `retype serve` or `retype serve --watch` will not build the project.

In your scenario, you would first need to run the command `retype build --watch`, then `retype serve --watch`.

The functionality of `--watch` is different depending on the command context:
- For `retype build --watch`, Retype will watch for file changes in your project `input` folder and rebuild the project if any changes are detected. The fresh build is automatically saved to the `output` folder. 
- For `retype serve --watch`, Retype will watch for any changes in the `output` folder. If any changes are detected, all clients (web browsers) that are viewing the website will be automatically updated. 

I will review the WARNING and ERROR messages for this scenario, as well as add some details to my descriptions in the matrix above.

**Reply by DannyBen** [reply]
This does not make any sense to me.

Why would anyone need to run `serve --watch`, if it does not also rebuild? It means that this command variation is only useful on the serve command if you run `retype build` in another terminal session.

This does not solve [my issue](https://github.com/retypeapp/retype/discussions/323#discussioncomment-3866513), as there does not seem to be a way to run a command that serves, watches and rebuilds, without opening the browser.

In my opinion, `--watch` should always be the same: Watch the source folder and rebuild on change.

---

Speaking of errors, I encountered another issue. When the port retype is configured on is in use:

1. `retype start` does not start the server and exits without any error.
2. `retype serve` fails with a non user friendly stack trace.

Seems like both should meet in the middle, and report a "Port 3000 already in use" error.

**Reply by geoffreymcgill** [reply]
An oversight on our part is the `retype build --watch && retype serve --watch` cannot be run in sequence, so yes, separate terminals would be required in that scenario. Far from ideal. 

I don't want to add **build** functionality to `retype serve` command. The `serve` command nicely focuses and encapsulates the functionality to _serve_ the generated website. Equally, the `build` command nicely focuses and encapsulates the functionality to _build_ the project.

I can see how using the `--watch` flag in both commands can cause confusion, so I'm going look into whether renaming to something else would be better. Using `--watch` with the `retype serve --watch` command is essentially instructing the web server to _live reload_ the clients. I'll put some time into this.

> This does not solve my issue, as there does not seem to be a way to run a command that serves, watches and rebuilds, without opening the browser.

The API is in a better place now, but agreed, it still does not cleanly solve the original issue of this thread.

So, we are proposing to add two new features to `v3.0`:

1. New `--no-open` flag with `-n` alias on `retype start`. The following demonstrates using the two flags:
```
retype start --no-open # or,
retype start -n
```

I see that the `--no-open` flag is used within similar projects to Retype, so looks like everyone has landed on the same solution.

2. New `start.open: false` project config. The following demonstrates setting the config in your projects **retype.yml** file.
```yml
start:
  open: false
```

If `open: false` is set in your **retype.yml** project config, then calling `retype start` would not open the web browser.

The `start.open` config could also be customized with the [`--override`](https://retype.com/guides/cli/#retype---override) flag. Obviously, this is not a super clean option, but it does allow for customization of the flags on an individual or per-call basis.

Hope this helps. Let me know what you think.

---

We are looking into the `Port 3000 already in use` errors and with some luck will have these fixed for the `v3.0` final release. Thanks for reporting.

**Reply by DannyBen** [reply]
> New `--no-open` flag with `-n` alias on retype start

Well... staying honest while knowing it might come off as abrasive - I think this whole thing got overcomplicated and over-engineered for no good reason.

I don't think this is the intuitive option. There is no reason to watch the target directory. The target only updates when the source updates.

Although having the option to disable the opening in the config file will help my use case, from UI perspective, it is not the place for it.
If I am sharing my repository with other users, some might want it to be different. Opening the browser is a CLI decision, not configuration decision. Even the same user might sometimes want to open and sometimes not.

> I see that the `--no-open` flag is used within similar projects to Retype, so looks like everyone has landed on the same solution.

Two wrongs don't make a right.

While I am obviously not in favor of how it is designed now, I can understand the "this is how it is now and we want to get on with it" argument.

**Reply by geoffreymcgill** [reply]
I appreciate the your comments and for being straightforward. I do agree with much of your comments.

>  I think this whole thing got overcomplicated and over-engineered for no good reason.

Over-engineered? what? of course not. We never over-engineer. /s

I kid, I kid.

The original API annoyed me too. I think it would be easy to argue that it was under-engineered at the time. The original API didn't have enough granularity, extensibility, and future-proofness to handle scenarios that we just didn't think about early on.

Basically we have only made two simple renamings of `retype watch` to `retype start` and `retype run` to `retype serve`, plus separated the `--watch` functionality into user configurable flags.

We also added missing functionality to choose if Retype should trigger the system default web browser to open. 

I see the benefits of the new API as:

1. 100% backwards compatible. Nothing will break if you upgrade your project to v3.0.
2. Super simple entry point. For 80%+ of users (probably more) the only thing you need to know is `retype start`.
3. Separation of Concerns. `build` is now responsible for building. `serve` is now responsible for serving.
4. more... I'll add here as I think of them

> There is no reason to watch the target directory.

There are a few. We have seen scenarios where post-processes outside of Retype update or add to the build output and those changes were not being reflected in the clients. That is not necessarily a Retype issue, but now it is supported. There are also features we are thinking about for future Retype releases that could take advantage of watching the `output` directory for changes.

Another reason is just that we needed to modularize how the server was live-updating the clients. Previously, the logic was co-mingled with the build process and lacked single responsibility. Now we have a nice separation of concerns and the `retype serve --watch` flag was mostly just a handy byproduct of that refactoring.

> from UI perspective, it is not the place for it.

Agreed. In your scenario, calling `retype start -n` would be the cleanest. The `start.open` project config is just another option that might be helpful to some users.

> Two wrongs don't make a right.

The situation is more nuanced than right or wrong. Unfortunately, there doesn't seem to perfect solution, or at least not yet.

With the new `-n` and `--no-open` flags and the `start.open` project config, there are clean options that should cover all scenarios.

The API can continue to evolve over time. None of this is forever. If better API ideas surface, we will remain flexible and open to make changes.

Hope this helps.

**Reply by geoffreymcgill** [reply]
We are making two changes on the `retype serve` cmd in Retype `v3.0.0`:

1. Removing the `--open` flag
2. Renaming `--watch` to `--live`

Here is the updated command matrix and descriptions:

| Command&nbsp;(v3.0)&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; | `init` | `build` | `serve` | `--watch` or `-w` |  `--live` or `-l` | `--no-open` or `-n` | 
|:------------------------------- |:------:|:-------:|:-------:|:---------:|:--------:| :--------:|
| `retype start` | ✅  | ✅   | ✅  | ✅    | ✅ |        | 
| `retype start -n` | ✅  | ✅   | ✅  | ✅    | ✅ | ✅   | 
| `retype init` | ✅  |       |       |         | |        | 
| `retype build` |      | ✅   |       |         | |        | 
| `retype build --watch` |      | ✅   |       | ✅    | |        | 
| `retype serve` |      |       | ✅   |         | |        | 
| `retype serve --live` |      |       | ✅   | | ✅    |        | 

| Command&nbsp;(v3.0)&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; | Description |
| ------------------------------- |  --- |
| `retype start`                | Build and serve the project using a local development only web server | 
| `retype start -n`           | Build and serve the project using a local development only web server, plus do not open a web browser. | 
| `retype init`                   | Initialize the project with a new `retype.yml` project file. |  
| `retype build`                | Generate a static website from the project. | 
| `retype build --watch` | Generate a static website from the project, watch for input file changes, and rebuild if any changes. | 
| `retype serve`               | Serve the website in a local development only web server. | 
| `retype serve --live` | Serve the website in a local development only web server and live reload open browsers when a change in the project output is detected. |

When the `retype start` command is called, if a `retype.yml` project config file is found, the `init` process is skipped. There is essentially no cost to `init` when running `retype start`.

**Reply by geoffreymcgill** [reply]
The `"Port 3000 already in use"` error has been fixed. See [comment](https://github.com/retypeapp/retype/discussions/323#discussioncomment-5988608). @DannyBen Thanks for reporting. 

Retype v3.0 should be released within the next 24 hours.

**Comment by DannyBen** [comment]
Any plans to incorporate this soon?
My workflow is:

1. Run `retype watch`
2. Be annoyed by the opened text browser
3. Press Q twice to quit the opened text browser
4. Relax
5. Start working
6. Think about how annoying that was
7. Relax
8. Go back to working

😝

**Reply by geoffreymcgill** [reply]
The commands [proposed](https://github.com/retypeapp/retype/discussions/323#discussioncomment-2876308) above will be available in Retype v3.0.

We will have a `3.0.0-preview` release available first, so you'll be able to test the new commands and any required changes can be made before the locking in the `3.0.0` final release. 

Hope this helps.

**Reply by geoffreymcgill** [reply]
> My workflow is:
> 
> 1. Run `retype watch`
> 2. Be annoyed by the opened text browser
> 3. Press Q twice to quit the opened text browser
> 4. Relax
> 5. Start working
> 6. Think about how annoying that was
> 7. Relax
> 8. Go back to working

New...

1. Run `retype start -n`
2. Start working

or, set:

```yml
start:
  open: false
```
then, 
1. Run `retype start`
2. Start working
